src/core/ee/emotion_breakpoint.cpp

```diff
@@ -0,0 +1,745 @@
+        }
+        case EERegisterKind::LO:
+        {
+            if(id != 0)
+                Errors::die("attempted to get name of invalid EE register");
+            return "lo";
+        }
+        case EERegisterKind::HI1:
+        {
+            if(id != 0)
+                Errors::die("attempted to get name of invalid EE register");
+            return "hi1";
+        }
+        case EERegisterKind::LO1:
+        {
+            if(id != 0)
+                Errors::die("attempted to get name of invalid EE register");
+            return "lo1";
+        }
+        default:
+            return "unknown register";
+    }
+}
+
+// get the register name (like v0, a0, ...)
+std::string IOPRegisterDefinition::get_name()
+{
+    if(id >= 32)
+        Errors::die("attempted to get name of invalid IOP register");
+    switch(kind)
+    {
+        case IOPRegisterKind::GPR:
+        {
+            return std::string(IOP::REG(id));
+        }
+        case IOPRegisterKind::HI:
+        {
+            if(id != 0)
+                Errors::die("attempted to get name of invalid IOP register");
+            return "hi*";
+        }
+        case IOPRegisterKind::LO:
+        {
+            if(id != 0)
+                Errors::die("attempted to get name of invalid IOP register");
+            return "lo*";
+        }
+        default:
+            return "unknown register";
+    }
+}
+
+
+
+std::string EmotionBreakpoints::PC::get_description()
+{
+    char buff[64];
+    sprintf(buff, "pc = 0x%0X", _addr);
+    return std::string(buff);
+}
+
+bool EmotionBreakpoints::PC::check(EmotionEngine *emu)
+{
+    return emu->get_PC() == _addr;
+}
+
+bool EmotionBreakpoints::PC::check(IOP *iop)
+{
+    return iop->get_PC() == _addr;
+}
+
+bool EmotionBreakpoints::PC::clear_after_hit()
+{
+    return _clear_after_hit;
+}
+
+template<typename T>
+std::string EmotionBreakpoints::Memory<T>::get_description()
+{
+    char c[128];
+    sprintf(c, "%s @ [0x%x] %s ", mem_compare_type_names[(uint64_t)_type], _addr, mem_compare_kind_names[(uint64_t)_kind]);
+    std::string result(c);
+    result += to_string_hex(_value);
+    return result;
+}
+
+namespace EmotionBreakpoints
+{
+    // uint128_t can't be "to_string"ed.
+    template <>
+    std::string Memory<uint128_t>::get_description()
+    {
+        char c[128];
+        sprintf(c, "%s @ [0x%x] %s 0x%lx%lx", mem_compare_type_names[(uint64_t)_type], _addr,
+                mem_compare_kind_names[(uint64_t)_kind], _value.hi, _value.lo);
+        std::string result(c);
+        return result;
+    }
+
+    // floats shouldn't have 0x...
+    template<>
+    std::string EmotionBreakpoints::Memory<float>::get_description()
+    {
+        char c[128];
+        sprintf(c, "%s @ [0x%x] %s ", mem_compare_type_names[(uint64_t)_type], _addr, mem_compare_kind_names[(uint64_t)_kind]);
+        std::string result(c);
+        result += std::to_string(_value);
+        return result;
+    }
+}
+
+template<typename T>
```

Comment:
It should be considered whether or not templates are needed in these EmotionBreakpoints functions. For example, instead of using a template + switch/case of size here, something like this can be implemented instead.

```cpp
bool EmotionBreakpoints::Memory<T>::check8(EmotionEngine *ee)
{
    return compare_memory(ee->read8(_addr), _value, _kind);
}

bool EmotionBreakpoints::Memory<T>::check16(EmotionEngine *ee)
{
    return compare_memory(ee->read16(_addr), _value, _kind);
}
```

etc.

Reply:
I originally had something like you said with memcmp, but this won't work easily with inequality comparisons.  Note that the check function is also a virtual function from the DebuggerBreakpoint class, and the caller of check has no idea what kind of breakpoint it actually is, so adding separate check8, check16, ... functions would require rewriting a lot of the other breakpoint code.